index.js

```diff
@@ -237,3 +241,34 @@ function arrObjKeys (obj, inspect) {
     }
     return xs;
 }
+
+// Returns the object's constructor name or null if it is a plain object
+// or doesn't have a prototype.
+function getTypeString(o) {
+    if (Object.prototype.toString(o) !== '[object Object]') return null;
+    var prototype = getPrototype(o);
+    if (!prototype) return null;
+
+    var constructorName = o.constructor ? o.constructor.name : null;
+    var isPlainObject = constructorName === 'Object' && looksLikeObjectPrototype(prototype);
+    if (isPlainObject) {
+      return null;
+    }
+
+    return constructorName;
+}
+
+// Indicates whether the specified object appears to be Object.prototype,
+// regardless of the object's realm.
+function looksLikeObjectPrototype(o) {
+  if (o === Object.prototype) return true;
+
+  // Cross-realm objects use a different Object, so we have to use a heuristic.
+  return !getPrototype(o)
+      && o.hasOwnProperty('hasOwnProperty')
+      && o.hasOwnProperty('isPrototypeOf')
+      && o.hasOwnProperty('propertyIsEnumerable')
+      && o.hasOwnProperty('toLocaleString')
```

Comment:
should these check that they're functions, and ideally, that they behave as expected?

Reply:
I don't think there's anything to be gained by that. Consider that you have an object with *own properties* `hasOwnProperty`, `isPrototypeOf`, etc. Either this *is* `Object.prototype`, or it's an object that's trying really hard to look like it. In the latter case, chances are that all those will be functions and that they'll behave correctly.

What's the worst that can happen? The worst case is that you have an object that's trying hard to look like a plain object, and that your library prints it just like a plain object. I don't think that's bad.

After all, this library isn't generating structured information. It's generating a debug string that is supposed to help a developer understand an object graph. And I don't think that anybody will complain if something that's virtually indistinguishable from a plain object is visualized as a plain object.

Reply:
That's a fair point. However, to be indistinguishable, *this library* would have to be unable to distinguish it.

One extreme length we could go to is ensuring not just that they're all functions, but also that `Function.prototype.toString.call` indicates it's native code, which would work cross-realm.